x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java

```diff
@@ -250,112 +298,165 @@ public void onFailure(Exception e) {
-        String originalIndexName,
+        String sourceIndexName,
         String tmpIndexName,
         String rollupIndexName,
         ActionListener<AcknowledgedResponse> listener
     ) {
-        // Update rollup metadata to include this index
+        // Update cluster state for the data stream to include the rollup index and exclude the source index
         clusterService.submitStateUpdateTask("update-rollup-metadata", new ClusterStateUpdateTask() {
-            @Override
-            public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
-                // Everything went well, time to delete the temporary index
-                deleteTmpIndex(originalIndexName, tmpIndexName, listener, null);
-            }
-
             @Override
             public ClusterState execute(ClusterState currentState) {
-                IndexMetadata rollupIndexMetadata = currentState.getMetadata().index(rollupIndexName);
-                Index rollupIndex = rollupIndexMetadata.getIndex();
-                IndexAbstraction originalIndex = currentState.getMetadata().getIndicesLookup().get(originalIndexName);
-
+                IndexAbstraction sourceIndex = currentState.getMetadata().getIndicesLookup().get(sourceIndexName);
                 Metadata.Builder metadataBuilder = Metadata.builder(currentState.metadata());
-                if (originalIndex.getParentDataStream() != null) {
+                if (sourceIndex.getParentDataStream() != null) {
+                    IndexMetadata rollupIndexMetadata = currentState.getMetadata().index(rollupIndexName);
+                    Index rollupIndex = rollupIndexMetadata.getIndex();
                     // If rolling up a backing index of a data stream, add rolled up index to backing data stream
-                    DataStream originalDataStream = originalIndex.getParentDataStream().getDataStream();
-                    List<Index> backingIndices = new ArrayList<>(originalDataStream.getIndices().size() + 1);
-                    // Adding rollup indices to the beginning of the list will prevent rollup indices from ever being
+                    DataStream originalDataStream = sourceIndex.getParentDataStream().getDataStream();
+                    List<Index> backingIndices = new ArrayList<>(originalDataStream.getIndices().size());
+                    // Adding the rollup index to the beginning of the list will prevent it from ever being
                     // considered a write index
                     backingIndices.add(rollupIndex);
-                    backingIndices.addAll(originalDataStream.getIndices());
+                    // Add all indices except the source index
+                    backingIndices.addAll(originalDataStream.getIndices().stream().filter(i -> i.getName() != sourceIndexName).toList());
                     DataStream dataStream = new DataStream(
                         originalDataStream.getName(),
                         backingIndices,
```

Comment:
Adding the rollup index to the data stream and remove the source index, hopefully in an atomic action.

@martijnvg your feedback on this would be greatly appreciated. Thanks

Reply:
The data stream in cluster state is updated in a atomic way here, since
the original index is removed and rolled up index is added in the same `ClusterStateUpdateTask` instance.

Reply:
Thank you. I pushed new code that uses `DataStream#replaceBackingIndex(...)`